examples/ui/directional_navigation.rs

```diff
@@ -0,0 +1,417 @@
+//!
+//! In this example, we will set up a simple UI with a grid of buttons that can be navigated using the arrow keys or gamepad input.
+
+use bevy::{
+    input_focus::{
+        directional_navigation::{
+            DirectionalNavigation, DirectionalNavigationMap, DirectionalNavigationPlugin,
+        },
+        InputDispatchPlugin, InputFocus, InputFocusVisible,
+    },
+    math::{CompassOctant, FloatOrd},
+    picking::{
+        backend::HitData,
+        pointer::{Location, PointerId},
+    },
+    prelude::*,
+    utils::{HashMap, HashSet},
+};
+use bevy_render::camera::NormalizedRenderTarget;
+
+fn main() {
+    App::new()
+        // Input focus is not enabled by default, so we need to add the corresponding plugins
+        .add_plugins((
+            DefaultPlugins,
+            InputDispatchPlugin,
+            DirectionalNavigationPlugin,
+        ))
+        // This resource is canonically used to track whether or not to render a focus indicator
+        // It starts as false, but we set it to true here as we would like to see the focus indicator
+        .insert_resource(InputFocusVisible(true))
+        // We've made a simple resource to keep track of the actions that are currently being pressed for this example
+        .init_resource::<ActionState>()
+        .add_systems(Startup, setup_ui)
+        // Input is generally handled during PreUpdate
+        // We're turning inputs into actions first, then using those actions to determine navigation
+        .add_systems(PreUpdate, (process_inputs, navigate).chain())
+        .add_systems(
+            Update,
+            (
+                // We need to show which button is currently focused
+                highlight_focused_element,
+                // Pressing the "Interact" button while we have a focused element should simulate a click
+                interact_with_focused_button,
+                // We're doing a tiny animation when the button is interacted with,
+                // so we need a timer and a polling mechanism to reset it
+                reset_button_after_interaction,
+            ),
+        )
+        // This observer is added globally, so it will respond to *any* trigger of the correct type.
+        // However, we're filtering in the observer's query to only respond to button presses
+        .add_observer(universal_button_click_behavior)
+        .run();
+}
+
+const NORMAL_BUTTON: Srgba = bevy::color::palettes::tailwind::BLUE_400;
+const PRESSED_BUTTON: Srgba = bevy::color::palettes::tailwind::BLUE_500;
+const FOCUSED_BORDER: Srgba = bevy::color::palettes::tailwind::BLUE_50;
+
+// This observer will be triggered whenever a button is pressed
+// In a real project, each button would also have its own unique behavior,
+// to capture the actual intent of the user
+fn universal_button_click_behavior(
+    // We're using a on-mouse-down trigger to improve responsiveness;
+    // Clicked is better when you want roll-off cancellation
```

Comment:
Might be nicer to have separate effects that get triggered on `Pressed` and `Clicked` rather than explain why it looks like we're doing the "wrong" thing.

But that could be a future improvement.